data/base.go

```diff
@@ -0,0 +1,286 @@
+	Type     string        `json:"type" bson:"type" valid:"required"`
+	UploadID string        `json:"uploadId" bson:"uploadId" valid:"-"`
+
+	//optional data
+	DeviceTime       string        `json:"deviceTime,omitempty" bson:"deviceTime,omitempty" valid:"omitempty,timestr"`
+	TimezoneOffset   int           `json:"timezoneOffset,omitempty" bson:"timezoneOffset,omitempty" valid:"omitempty,timezoneoffset"`
+	ConversionOffset int           `json:"conversionOffset,omitempty" bson:"conversionOffset,omitempty" valid:"omitempty,required"`
+	ClockDriftOffset int           `json:"clockDriftOffset,omitempty" bson:"clockDriftOffset,omitempty" valid:"omitempty,required"`
+	Payload          interface{}   `json:"payload,omitempty" bson:"payload,omitempty" valid:"omitempty,payload"`
+	Annotations      []interface{} `json:"annotations,omitempty" bson:"annotations,omitempty" valid:"omitempty,annotations"`
+
+	//used for versioning and de-deping
+	Internal `bson:",inline"`
+}
+
+//Internal are existing fields used for versioning and de-deping
+type Internal struct {
+	CreatedTime   string `json:"createdTime" bson:"createdTime" valid:"timestr"`
+	GroupID       string `json:"-" bson:"_groupId" valid:"required"`
+	ActiveFlag    bool   `json:"-" bson:"_active" valid:"required"`
+	SchemaVersion int    `json:"-" bson:"_schemaVersion" valid:"required,min=0"`
+	Version       int    `json:"-" bson:"_version,omitempty" valid:"-"`
+}
+
+var (
+	//InternalFields are what we only use internally in the service and don't wish to return
+	InternalFields = map[string]interface{}{
+		"_groupId":       0,
+		"_active":        0,
+		"_schemaVersion": 0,
+		"_version":       0,
+		"createdTime":    0,
+	}
+)
+
+const (
+	//UserIDField is the userID
+	UserIDField = "userId"
+	//GroupIDField id the groupID
+	GroupIDField = "groupId"
+
+	deviceIDField = "deviceId"
+	uploadIDField = "uploadId"
+
+	timezoneOffsetField   = "timezoneOffset"
+	conversionOffsetField = "conversionOffset"
+	clockDriftOffsetField = "clockDriftOffset"
+
+	typeField       = "type"
+	timeField       = "time"
+	deviceTimeField = "deviceTime"
+
+	payloadField     = "payload"
+	annotationsField = "annotations"
+
+	//validation
+	tzValidationLowerLimit = -840
+	tzValidationUpperLimit = 720
+	timeStrValidationMsg   = "Times need to be ISO 8601 format and not in the future"
+
+	timeStrTag        validate.ValidationTag = "timestr"
+	timeObjTag        validate.ValidationTag = "timeobj"
+	timeZoneOffsetTag validate.ValidationTag = "timezoneoffset"
+	payloadTag        validate.ValidationTag = "payload"
+	annotationsTag    validate.ValidationTag = "annotations"
+)
+
+//BuildBase builds the base fields that all device data records contain
+func BuildBase(obj map[string]interface{}) (Base, *Error) {
+
+	errs := NewError(obj)
+	cast := NewCaster(errs)
+
+	base := Base{
+		_ID:      bson.NewObjectId(),
+		ID:       bson.NewObjectId().Hex(),
+		UserID:   cast.ToString(UserIDField, obj[UserIDField]),
```

Comment:
Perhaps consider a caster that pulls out of the object AND casts.

With pulling out using `obj[UserIDField]` you turn a missing field into an empty string, which are not the same thing.